### DISCUSSION TITLE
Modular web app

### DISCUSSION THREAD

**Discussion by majidbl** [actix/actix-web#1977] [Q&A]
I wanna have async pub trait for handler like
``` rust
pub async trait handler{
    async fn simple_handler(&self, req) -> Response;
}
```
How can I do that?

**Comment by fakeshadow** [comment]
There is no async trait in Rust. You can use normal trait to return a boxed async block but it comes with an overhead.
On the other hand async functions as handler are zero overhead.

**Reply by majidbl** [reply]
thanks for your answer.
but I need implement clean architecture in my project.
how can i ensure that all handler implemented and how i can pass an arguments to handler function.

**Reply by fakeshadow** [reply]
handler functions are used as a closure. You can use the move keyword to capture the outter variables.
Something like this would work.
```rust
#[actix_web::main]
async fn main() -> std::io::Result<()> {

    let state = std::sync::Arc::new(123u32);

    HttpServer::new(move || {
        let state = state.clone();
        App::new().service(web::resource("/").to(move || {
            let state = state.clone();
            async move { format!("state is {}", &*state) }
        }))
    })
    .bind("127.0.0.1:8080")?
    .run()
    .await
}
```
And you can see this would be very messy when your project goes up in size that would have multiple vars and/or handlers. So this is not what you want to do normally in actix-web.

In actix-web there is `App::data` where you can pass your variables and states to the whole App and access them in handler functions. The clean way would be you make your states and variables structs that passed around with `App::data` and impl traits and/or methods for these structs for access and mutating them.

**Reply by majidbl** [reply]
thanks.
truly this function was useful.
i think a function wrapper for convert function returns to Response should be needed

**Reply by spf5000** [reply]
Looking at the code, I can already see the Handler trait that is used in the web module, but that trait isn’t public in the library which I don’t fully understand. Skimming the code, it looks like this trait in already wrapped into an async function using the Route.to() function via magic I don’t fully understand yet. I’m new to Rust, let alone actix, so happy to be told why this is foolish/hard. 

That said, I agree with the original poster here about wanting a handler trait. The idea of having to store all of my dependencies in the Application state feels odd to me. Creating my own structs (implementing the handler trait) with the needed dependencies and passing those structs into the App/Server feels cleaner/easier to me.

**Reply by spf5000** [reply]
Looks like this is actually public under dev in 4.0 :)

https://github.com/actix/actix-web/blob/web-v4.0.0-beta.1/src/lib.rs#L124
